### PR TITLE
[CU-3p0r4ry] rpc for fe assets

### PR DESCRIPTION
**Pull request by andresmechali**
## Issue
https://app.clickup.com/t/3p0r4ry

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-3p0r4ry RPC/API for Front End Assets](https://app.clickup.com/t/3p0r4ry)

File: code/parachain/frame/assets-registry/src/lib.rs [hunk 1]
```diff
@@ -309,6 +315,24 @@ pub mod pallet {
...
+			LocalToForeign::<T>::iter()
+				.map(|(local_id, foreign_metadata)| {
+					let decimals = match foreign_metadata.decimals {
+						Some(exponent) => exponent,
+						_ => 12_u32,
+					};
+
+					Asset {
+						name: None,
+						id: local_id.into() as u64,
```
Comment (benluelo) [inline code/parachain/frame/assets-registry/src/lib.rs:329]: .into() followed by `as` is a bit strange, using .into() only should be enough

File: code/parachain/frame/assets/src/lib.rs [hunk 2]
```diff
@@ -116,6 +118,15 @@ pub mod pallet {
 		#[pallet::constant]
 		type NativeAssetId: Get<Self::AssetId>;
 		type GenerateCurrencyId: CurrencyFactory<AssetId = Self::AssetId, Balance = Self::Balance>;
+		/// Identifier for the class of foreign asset.
+		type ForeignAssetId: FullCodec
```
Comment (benluelo) [inline code/parachain/frame/assets/src/lib.rs:122]: Is this used anywhere in the pallet?

File: code/parachain/runtime/dali/src/lib.rs [hunk 3]
```diff
@@ -1455,13 +1457,17 @@ impl_runtime_apis! {
...
 		}
 
-		fn list_assets() -> Vec<Asset> {
-			CurrencyId::list_assets()
+		fn list_assets() -> Vec<Asset<ForeignAssetId>> {
+			let mut assets = CurrencyId::list_assets();
+			let mut foreign_assets = assets_registry::Pallet::<Runtime>::get_foreign_assets_list();
+			assets.append(&mut foreign_assets);
+
+			assets
```
Comment (benluelo) [inline code/parachain/runtime/dali/src/lib.rs:1470]: This needs to take into account assets that may be in both

Comment (benluelo) [reply code/parachain/runtime/dali/src/lib.rs:1470]: (otherwise there will be duplicated assets)

File: code/parachain/frame/assets-registry/src/lib.rs (hunk 1 above)
Comment (andresmechali) [reply code/parachain/frame/assets-registry/src/lib.rs:329]: That's what I though but it's not. `local_id` is a `LocalAssetId` and with `.into()` it turns into a `u128`

File: code/parachain/runtime/dali/src/lib.rs (hunk 3 above)
Comment (andresmechali) [reply code/parachain/runtime/dali/src/lib.rs:1470]: Makes sense. But how can we know? Local ids would be different anyway no?

File: code/parachain/frame/assets-registry/src/lib.rs (hunk 1 above)
Comment (benluelo) [reply code/parachain/frame/assets-registry/src/lib.rs:329]: then you should use `.try_into()` to go from `u128` to `u64`, or change the `Asset.id` type to `u128` - `as` will silently truncate the top 64 bits of `u128` when converting from `u64`

File: code/parachain/frame/assets-registry/src/lib.rs [hunk 4]
```diff
@@ -91,8 +92,13 @@ pub mod pallet {
 	/// Mapping local asset to foreign asset.
 	#[pallet::storage]
 	#[pallet::getter(fn from_local_asset)]
-	pub type LocalToForeign<T: Config> =
-		StorageMap<_, Twox128, T::LocalAssetId, ForeignMetadata<T::ForeignAssetId>, OptionQuery>;
+	pub type LocalToForeign<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
```
Comment (benluelo) [inline code/parachain/frame/assets-registry/src/lib.rs:97]: why was this changed from `Twox128`?

File: code/parachain/frame/composable-traits/src/assets.rs [hunk 5]
```diff
@@ -25,7 +26,9 @@ impl BasicAssetMetadata {
 
 #[derive(Decode, Encode, Debug, Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub struct Asset {
-	pub name: Vec<u8>,
+pub struct Asset<ForeignId> {
+	pub name: Option<Vec<u8>>,
 	pub id: u64,
```
Comment (benluelo) [inline code/parachain/frame/composable-traits/src/assets.rs:31]: This should probably be a generic field instead of a concrete `u64`

File: code/parachain/frame/assets-registry/src/lib.rs (hunk 1 above)
Comment (andresmechali) [reply code/parachain/frame/assets-registry/src/lib.rs]: changed to u128 👌🏼

File: code/parachain/frame/assets/src/lib.rs (hunk 2 above)
Comment (andresmechali) [reply code/parachain/frame/assets/src/lib.rs]: Actually no, I added it at some point and then removed it. Well spotted.

File: code/parachain/frame/assets-registry/src/lib.rs (hunk 4 above)
Comment (andresmechali) [reply code/parachain/frame/assets-registry/src/lib.rs:97]: `Twox128` doesn't allow me to `iter()` as `ReversibleStorageHasher` is not satisfied.

File: code/parachain/frame/composable-traits/src/assets.rs (hunk 5 above)
Comment (andresmechali) [reply code/parachain/frame/composable-traits/src/assets.rs]: We should be ok with a `u128` no?

File: code/parachain/runtime/composable/src/lib.rs [hunk 6]
```diff
@@ -813,6 +813,7 @@ construct_runtime!(
 		CrowdloanRewards: crowdloan_rewards = 56,
 		Assets: assets = 57,
 		GovernanceRegistry: governance_registry = 58,
+		// AssetsRegistry: assets_registry = 59,
```
Comment (dzmitry-lahoda) [inline code/parachain/runtime/composable/src/lib.rs:816]: why commented out?

File: code/parachain/runtime/primitives/src/currency.rs [hunk 7]
```diff
@@ -89,9 +89,9 @@ macro_rules! list_assets {
 			}
 		}
 
-		pub fn list_assets() -> Vec<Asset> {
+		pub fn list_assets() -> Vec<Asset<XcmAssetLocation>> {
 			[
-				$(Asset { id: CurrencyId::$NAME.0 as u64, name: stringify!($NAME).as_bytes().to_vec() },)*
+				$(Asset { id: CurrencyId::$NAME.0 as u128, name: Some(stringify!($NAME).as_bytes().to_vec()), decimals: 12_u8.into(), foreign_id: None },)*
```
Comment (dzmitry-lahoda) [inline code/parachain/runtime/primitives/src/currency.rs:94]: KSM and USD* have foreign asset id too or will have as soon as registered.

Comment (dzmitry-lahoda) [reply code/parachain/runtime/primitives/src/currency.rs:94]: so it  there should be join of registry with hard coded

File: code/parachain/frame/assets-registry/src/lib.rs (hunk 4 above)
Comment (dzmitry-lahoda) [reply code/parachain/frame/assets-registry/src/lib.rs:97]: alocal asset id is permissioed, so i  can be 64 bit hash with concat. i am sure there is such. in other examples.

Comment (dzmitry-lahoda) [reply code/parachain/frame/assets-registry/src/lib.rs:97]: twox128 was overkill

Comment (dzmitry-lahoda) [reply code/parachain/frame/assets-registry/src/lib.rs:97]: you could iterate values and get local if by reverse map. foreign to local and local to foreign

Comment (dzmitry-lahoda) [reply code/parachain/frame/assets-registry/src/lib.rs:97]: we have some assets on rococo, may be on picasso. not big deal now. but recreating them is gated by multisig and democracy.

File: code/parachain/runtime/composable/src/lib.rs (hunk 6 above)
Comment (andresmechali) [reply code/parachain/runtime/composable/src/lib.rs:816]: Forgot to remove it

File: code/parachain/frame/assets-registry/src/lib.rs (hunk 4 above)
Comment (andresmechali) [reply code/parachain/frame/assets-registry/src/lib.rs:97]: `Twox64Concat` works as well, but still has the problem of changing storage. Should I use that, or go for the `ForeignToLocal-LocalToForeign` double iteration?

Comment (benluelo) [reply code/parachain/frame/assets-registry/src/lib.rs:97]: RPCs have no on-chain cost, so it should be fine to do the double iteration

File: code/parachain/frame/composable-traits/src/assets.rs (hunk 5 above)
Comment (benluelo) [reply code/parachain/frame/composable-traits/src/assets.rs]: Yes, but it would probably be best to make it generic. This can be addressed in another ticket though, feel free to resolve :+1:

File: code/parachain/runtime/dali/src/lib.rs (hunk 3 above)
Comment (benluelo) [reply code/parachain/runtime/dali/src/lib.rs:1469]: @dzmitry-lahoda left a comment relating to this as well, I think he may have more insight into how to properly handle this than I do

File: code/parachain/runtime/primitives/src/currency.rs (hunk 7 above)
Comment (andresmechali) [reply code/parachain/runtime/primitives/src/currency.rs:94]: @dzmitry-lahoda  What can we use to identify already existing assets, given that when we register we are not giving any input on `local_id` or `name` or `symbol`?

File: code/parachain/frame/composable-traits/src/assets.rs (hunk 5 above)
Comment (andresmechali) [reply code/parachain/frame/composable-traits/src/assets.rs]: https://app.clickup.com/t/3mzpfw2

File: code/parachain/runtime/primitives/src/currency.rs (hunk 7 above)
Comment (dzmitry-lahoda) [reply code/parachain/runtime/primitives/src/currency.rs:94]: Encoding works like that https://drive.google.com/file/d/169Jb6bdJKIGtK0bVqYlQgDSvN0nSfNhq/view  and can see some examples https://github.com/ComposableFi/composable/pull/2035#issuecomment-1288758296

Comment (dzmitry-lahoda) [reply code/parachain/runtime/primitives/src/currency.rs:94]: will update after check

Comment (andresmechali) [reply code/parachain/runtime/primitives/src/currency.rs:94]: Thanks @dzmitry-lahoda , it helped me understanding better xcm.
Can you provide me with a concrete example on how a `MultiLocation` would look like when we register on our chain for example `USDT` (which locally has id `130`) , which let's assume has id `111` on some parachain with id `222`?

File: code/parachain/runtime/dali/src/lib.rs (hunk 3 above)
Comment (andresmechali) [reply code/parachain/runtime/dali/src/lib.rs:1469]: @benluelo as far as I understand, there will be no duplicated assets. Registering an asset requires sudo, and nobody will register one that is already hardcoded.

**Review by dzmitry-lahoda — APPROVED** [review]

File: code/parachain/runtime/dali/src/lib.rs (hunk 3 above)
Comment (dzmitry-lahoda) [reply code/parachain/runtime/dali/src/lib.rs:1469]: these are hardcode in PR for XCM. I can fix this RPC if you will at least hardcode KSM = (1,Here).into()

**Review by vimukthi-git — APPROVED** [review]
LGTM

**Comment by andresmechali** [comment]
@mergify refresh

**Comment by andresmechali** [comment]
@mergify refresh

**Review by easteregg — APPROVED** [review]

**Review by saadjhk — APPROVED** [review]

**Comment by andresmechali** [comment]
@mergify refresh

**Review by hussein-aitlahcen — APPROVED** [review]

File: code/parachain/runtime/primitives/src/currency.rs [hunk 8]
```diff
@@ -89,9 +89,24 @@ macro_rules! list_assets {
 			}
 		}
 
-		pub fn list_assets() -> Vec<Asset> {
+		const RELAY_NATIVE_CURRENCIES: [CurrencyId; 1] = [CurrencyId::KSM];
```
Comment (benluelo) [inline code/parachain/runtime/primitives/src/currency.rs:92]: Does this (and known_location) need to be in the macro?

**Review by benluelo — APPROVED** [review]

File: code/parachain/runtime/primitives/src/currency.rs (hunk 8 above)
Comment (andresmechali) [reply code/parachain/runtime/primitives/src/currency.rs:92]: Not exactly, it made sense for me as it was only used there but it can be extracted if it makes more sense

**Comment by andresmechali** [comment]
@mergify refresh

File: code/parachain/runtime/primitives/src/currency.rs (hunk 8 above)
Comment (andresmechali) [reply code/parachain/runtime/primitives/src/currency.rs:92]: I will resolve this to get it merged asap and can extract it later on some other PR
